<a href="https://colab.research.google.com/github/palashbaranwal/END3/blob/main/Session4/Sentiment_Analysis_Using_RNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Sentiment Analysis of IMDB dataset using RNN

#Using RNN

1. `Field` - One of the main concepts in TorchText. 
2. `Field` defines how our data should be processed. 
3. In our sentiment classification task the data consists of both the raw string of the review and the sentiment, either "pos" or "neg".
4. The parameters of a `Field` specify how the data should be processed.
5. We use the `TEXT` field to define how the review should be processed, and the `LABEL` field to process the sentiment.
6. Our `TEXT` field has tokenize='spacy' as an argument. This defines that the "tokenization" (the act of splitting the string into discrete "tokens") should be done using the spaCy tokenizer. If no tokenize argument is passed, the default is simply splitting the string on spaces. We also need to specify a tokenizer_language which tells torchtext which spaCy model to use. We use the en_core_web_sm model which has to be downloaded with python -m spacy download en_core_web_sm before you run this notebook!
7. LABEL is defined by a LabelField, a special subset of the Field class specifically used for handling labels. 
8. We also add `SEED` for reproducibility.

In [1]:
import torch
from torchtext.legacy import data

SEED = 1234

# https://pytorch.org/docs/stable/generated/torch.manual_seed.html 
# This sets the seed for generating random numbers.
torch.manual_seed(SEED)

# https://blog.titanwolf.in/a?ID=01700-1af15100-d8a4-4493-a0c7-2f7fbe09da96
# If this flag is set to True, the convolution algorithm returned each time will be determined, that is, the default algorithm
torch.backends.cudnn.deterministic = True

# https://torchtext.readthedocs.io/en/latest/data.html#field 
# Defines a datatype together with instructions for converting to Tensor. 
TEXT = data.Field(tokenize = 'spacy',
                  tokenizer_language = 'en_core_web_sm')


LABEL = data.LabelField(dtype = torch.float)

Another handy feature of TorchText is that it has `support for common datasets` used in natural language processing (NLP).

The following code automatically downloads the IMDb dataset and splits it into the canonical train/test splits as torchtext.datasets objects. It process the data using the Fields we have previously defined. The IMDb dataset consists of 50,000 movie reviews, each marked as being a positive or negative review.

In [2]:
from torchtext.legacy import datasets

train_data, test_data = datasets.IMDB.splits(TEXT, LABEL)

downloading aclImdb_v1.tar.gz


100%|██████████| 84.1M/84.1M [00:02<00:00, 30.4MB/s]


In [3]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 25000
Number of testing examples: 25000


In [4]:
# https://www.geeksforgeeks.org/vars-function-python/
# Returns __dict__ attribute for a module, class or instance
print(vars(train_data.examples[0]))

{'text': ['This', 'is', 'one', 'of', 'the', 'best', 'movies', 'I', "'ve", 'ever', 'seen', '.', 'It', 'has', 'very', 'good', 'acting', 'by', 'Hanks', ',', 'Newman', ',', 'and', 'everyone', 'else', '.', 'Definitely', 'Jude', 'Law', "'s", 'best', 'performance', '.', 'The', 'cinematography', 'is', 'excellent', ',', 'the', 'editing', 'is', 'about', 'as', 'good', ',', 'and', 'includes', 'a', 'great', 'original', 'score', 'that', 'really', 'fits', 'in', 'with', 'the', 'mood', 'of', 'the', 'movie', '.', 'The', 'production', 'design', 'is', 'also', 'a', 'factor', 'in', 'what', 'makes', 'this', 'movie', 'special', '.', 'To', 'me', ',', 'it', 'takes', 'a', 'lot', 'to', 'beat', 'Godfather', ',', 'but', 'the', 'fantastic', 'cinematography', 'displayed', 'wins', 'this', 'contest', '.', 'Definitely', 'a', 'Best', 'Picture', 'nominee', 'in', 'my', 'book', '.'], 'label': 'pos'}


The IMDb dataset only has train/test splits, so we need to create a validation set. We can do this with the `.split()` method.

By default this splits 70/30, however by passing a `split_ratio` argument, we can change the ratio of the split, i.e. a split_ratio of 0.8 would mean 80% of the examples make up the training set and 20% make up the validation set.

We also pass our random seed to the `random_state` argument, ensuring that we get the same train/validation split each time.

In [5]:
import random

train_data, valid_data = train_data.split(random_state = random.seed(SEED))

In [6]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of validation examples: {len(valid_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 17500
Number of validation examples: 7500
Number of testing examples: 25000


1. Next, we have to build a vocabulary. This is a effectively a look up table where every unique word in your data set has a corresponding index (an integer).

2. We do this as our machine learning model cannot operate on strings, only numbers. Each index is used to construct a one-hot vector for each word. A one-hot vector is a vector where all of the elements are 0, except one, which is 1, and dimensionality is the total number of unique words in your vocabulary, commonly denoted by **V**.

3. The number of unique words in our training set is over 100,000, which means that our one-hot vectors will have over 100,000 dimensions! This will make training slow and possibly won't fit onto your GPU (if you're using one).

4. There are two ways effectively cut down our vocabulary, we can either only take the top  most common words or ignore words that appear less than  times. We'll do the former, only keeping the top 25,000 words.

5. What do we do with words that appear in examples but we have cut from the vocabulary? We replace them with a special unknown or `<unk>` token. For example, if the sentence was "This film is great and I love it" but the word "love" was not in the vocabulary, it would become "This film is great and I `<unk>` it".

The following builds the vocabulary, only keeping the most common `max_size` tokens.

In [7]:
MAX_VOCAB_SIZE = 25_000

# Constructs a vocab for this field from one or more dataset
TEXT.build_vocab(train_data, max_size = MAX_VOCAB_SIZE)
LABEL.build_vocab(train_data)

**Why do we only build the vocabulary on the training set?** 

When testing any machine learning system you do not want to look at the test set in any way. We do not include the validation set as we want it to reflect the test set as much as possible.

In [8]:
print(f"Unique tokens in TEXT vocabulary: {len(TEXT.vocab)}")
print(f"Unique tokens in LABEL vocabulary: {len(LABEL.vocab)}")

Unique tokens in TEXT vocabulary: 25002
Unique tokens in LABEL vocabulary: 2


1. The two extra tokens in TEXT vocabulary are PAD and UNKnown.

2. The two tokens in LABEL vocabulary are text and label.

**Why is the vocab size 25002 and not 25000? One of the addition tokens is the `<unk>` token and the other is a `<pad>` token.**

When we feed sentences into our model, we feed a batch of them at a time, i.e. more than one at a time, and all sentences in the batch need to be the same size. Thus, to ensure each sentence in the batch is the same size, any shorter than the longest within the batch are padded.

In [9]:
print(TEXT.vocab.freqs.most_common(20))

[('the', 201539), (',', 191653), ('.', 165524), ('and', 109353), ('a', 108956), ('of', 100466), ('to', 93204), ('is', 76649), ('in', 61051), ('I', 54537), ('it', 53571), ('that', 48912), ('"', 44010), ("'s", 42865), ('this', 42419), ('-', 36845), ('/><br', 35299), ('was', 34783), ('as', 30307), ('movie', 29927)]


We can also see the vocabulary directly using either the `stoi` (string to int) or `itos` (int to string) method.

In [10]:
print(TEXT.vocab.itos[:10])

['<unk>', '<pad>', 'the', ',', '.', 'and', 'a', 'of', 'to', 'is']


We can also check the labels, ensuring 0 is for negative and 1 is for positive.

In [11]:
print(LABEL.vocab.stoi)

defaultdict(None, {'neg': 0, 'pos': 1})


The final step of preparing the data is creating the iterators. We iterate over these in the training/evaluation loop, and they return a batch of examples (indexed and converted into tensors) at each iteration.

We'll use a BucketIterator which is a special type of iterator that will return a batch of examples where each example is of a similar length, minimizing the amount of padding per example.

We also want to place the tensors returned by the iterator on the GPU (if you're using one). PyTorch handles this using torch.device, we then pass this device to the iterator.

In [19]:
BATCH_SIZE = 64

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

train_iterator, valid_iterator, test_iterator = data.BucketIterator.splits(
    (train_data, valid_data, test_data), 
    batch_size = BATCH_SIZE,
    device = device)

#### Build the Model

The next stage is building the model that we'll eventually train and evaluate.

There is a small amount of boilerplate code when creating models in PyTorch, note how our `RNN class is a sub-class of nn.Module` and the use of `super`.

Within the __init__ we define the layers of the module. Our three layers are an 

1. embedding layer, 
2. our RNN, and 
3. a linear layer. 

All layers have their parameters initialized to random values, unless explicitly specified.

The **embedding layer** is used to transform our sparse one-hot vector (sparse as most of the elements are 0) into a dense embedding vector (dense as the dimensionality is a lot smaller and all the elements are real numbers). This embedding layer is simply a single fully connected layer. As well as reducing the dimensionality of the input to the RNN, there is the theory that words which have similar impact on the sentiment of the review are mapped close together in this dense vector space. For more information about word embeddings, see here - https://monkeylearn.com/blog/word-embeddings-transform-text-numbers/.

The RNN layer is our RNN which takes in our dense vector and the previous hidden state h(t-1), which it uses to calculate the next hidden state, h(t). Finally, the linear layer takes the final hidden state and feeds it through a fully connected layer, f(h(t)) , transforming it to the correct output dimension.

#### Defining the forward function

1. The `forward` method is called when we feed examples into our model.
2. Each batch, `text`, is a tensor of size [sentence length, batch size]. That is a batch of sentences, each having each word converted into a one-hot vector.
3. You may notice that this tensor should have another dimension due to the one-hot vectors, however PyTorch conveniently stores a one-hot vector as it's index value, i.e. the tensor representing a sentence is just a tensor of the indexes for each token in that sentence. The act of converting a list of tokens into a list of indexes is commonly called `numericalizing`.
4. The input batch is then passed through the embedding layer to get `embedded`, which gives us a dense vector representation of our sentences. `embedded` is a tensor of size [sentence length, batch size, embedding dim]. 
5. `embedded` is then fed into the RNN. In some frameworks you must feed the initial hidden state, h0, into the RNN, however in PyTorch, if no initial hidden state is passed as an argument it defaults to a tensor of all zeros.
6. The **RNN returns 2 tensors**, 
    
    a. `output` of size [sentence length, batch size, hidden dim] and 
    
    b. `hidden` of size [1, batch size, hidden dim]. 

    c. `output` is the concatenation of the hidden state from every time step, whereas `hidden` is simply the final hidden state. 

We verify this using the `assert` statement. Note the `squeeze` method, which is used to remove a dimension of size 1.

**Finally**, we feed the last hidden state, hidden, through the linear layer, fc, to produce a prediction.

In [12]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
        
        super().__init__()
        
        self.embedding = nn.Embedding(input_dim, embedding_dim)
        
        self.rnn = nn.RNN(embedding_dim, hidden_dim)
        
        self.fc = nn.Linear(hidden_dim, output_dim)
        
    def forward(self, text):

        #text = [sent len, batch size]
        
        embedded = self.embedding(text)
        
        #embedded = [sent len, batch size, emb dim]
        
        output, hidden = self.rnn(embedded)
        
        #output = [sent len, batch size, hid dim]
        #hidden = [1, batch size, hid dim]
        
        assert torch.equal(output[-1,:,:], hidden.squeeze(0))
        
        return self.fc(hidden.squeeze(0))

We now create an instance of our RNN class.

1. The `input dimension` is the dimension of the one-hot vectors, which is equal to the vocabulary size.
2. The `embedding dimension` is the size of the dense word vectors. This is usually around 50-250 dimensions, but depends on the size of the vocabulary.
3. The `hidden dimension` is the size of the hidden states. This is usually around 100-500 dimensions, but also depends on factors such as on the vocabulary size, the size of the dense vectors and the complexity of the task.
4. The `output dimension` is usually the number of classes, however in the case of only 2 classes the output value is between 0 and 1 and thus can be 1-dimensional, i.e. a single scalar real number.

In [13]:

INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 100
HIDDEN_DIM = 256
OUTPUT_DIM = 1

model = RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)

Let's also create a function that will tell us how many trainable parameters our model has so we can compare the number of parameters across different models.

In [14]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f'The model has {count_parameters(model):,} trainable parameters')

The model has 2,592,105 trainable parameters


####Training the model

Now we will set up the training and then train the model.

First, we'll create an optimizer. This is the algorithm we use to update the parameters of the module. Here, we'll use stochastic gradient descent (SGD). 
1. The first argument is the parameters will be updated by the optimizer, 
2. The second is the learning rate, i.e. how much we'll change the parameters by when we do a parameter update.

In [15]:
import torch.optim as optim

optimizer = optim.SGD(model.parameters(), lr=1e-3)

Next, we'll **define our loss function**. In PyTorch this is commonly called a **criterion**.

1. The loss function here is `binary cross entropy with logits`.
2. Our model currently outputs an unbound real number. As our labels are either 0 or 1, we want to restrict the predictions to a number between 0 and 1. We do this using the **sigmoid or logit functions**.
3. We then use this bound scalar to calculate the loss using binary cross entropy.
4. The **BCEWithLogitsLoss** criterion carries out both the sigmoid and the binary cross entropy steps.

In [16]:
criterion = nn.BCEWithLogitsLoss()

Using `.to`, we can place the model and the criterion on the GPU (if we have one).



In [17]:
model = model.to(device)
criterion = criterion.to(device)

NameError: ignored

Our **criterion function** calculates the loss, however we have to write our function to calculate the accuracy.

1. This function first feeds the predictions through a sigmoid layer, squashing the values between 0 and 1, we then round them to the nearest integer. This rounds any value greater than 0.5 to 1 (a positive sentiment) and the rest to 0 (a negative sentiment).
2. We then calculate how many rounded predictions equal the actual labels and average it across the batch.

In [18]:
def binary_accuracy(preds, y):
    """
    Returns accuracy per batch, i.e. if you get 8/10 right, this returns 0.8, NOT 8
    """
    #round predictions to the closest integer
    rounded_preds = torch.round(torch.sigmoid(preds))
    correct = (rounded_preds == y).float() #convert into float for division 
    acc = correct.sum() / len(correct)
    return acc

####The Train Function

1. The train function iterates over all examples, one batch at a time.
2. **model.train()** is used to put the model in "training mode", which turns on dropout and batch normalization. Although we aren't using them in this model, it's good practice to include it.
3. For each batch, we first zero the gradients. Each parameter in a model has a `grad` attribute which stores the gradient calculated by the criterion. PyTorch does not automatically remove (or "zero") the gradients calculated from the last gradient calculation, so they must be manually zeroed.

4. We then feed the batch of sentences, `batch.text`, into the model. Note, you do not need to do model.forward(batch.text), simply calling the model works. The `squeeze` is needed as the predictions are initially size [batch size, 1], and we need to remove the dimension of size 1 as PyTorch expects the predictions input to our criterion function to be of size [batch size].

5. The loss and accuracy are then calculated using our predictions and the labels, batch.label, with the loss being averaged over all examples in the batch.

6. We calculate the gradient of each parameter with loss.backward(), and then update the parameters using the gradients and optimizer algorithm with optimizer.step().

6. The loss and accuracy is accumulated across the epoch, the .item() method is used to extract a scalar from a tensor which only contains a single value.

7. Finally, we return the loss and accuracy, averaged across the epoch. The len of an iterator is the number of batches in the iterator.

**You may recall when initializing the LABEL field, we set dtype=torch.float. This is because TorchText sets tensors to be LongTensors by default, however our criterion expects both inputs to be FloatTensors. Setting the dtype to be torch.float, did this for us. The alternative method of doing this would be to do the conversion inside the train function by passing batch.label.float() instad of batch.label to the criterion.*

In [ ]:
def train(model, iterator, optimizer, criterion):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.train()
    
    for batch in iterator:
        
        optimizer.zero_grad()
                
        predictions = model(batch.text).squeeze(1)
        
        loss = criterion(predictions, batch.label)
        
        acc = binary_accuracy(predictions, batch.label)
        
        loss.backward()
        
        optimizer.step()
        
        epoch_loss += loss.item()
        epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)

####The Evaluate Function

1. `evaluate` is similar to `train`, with a few modifications as you don't want to update the parameters when evaluating.

2. `model.eval()` puts the model in "evaluation mode", this turns off dropout and batch normalization. Again, we are not using them in this model, but it is good practice to include them.
3. No gradients are calculated on PyTorch operations inside the with `no_grad()` block. This causes less memory to be used and speeds up computation.
4. The rest of the function is the same as train, with the removal of `optimizer.zero_grad(), loss.backward() and optimizer.step()`, as we do not update the model's parameters when evaluating.

In [ ]:
def evaluate(model, iterator, criterion):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.eval()
    
    with torch.no_grad():
    
        for batch in iterator:

            predictions = model(batch.text).squeeze(1)
            
            loss = criterion(predictions, batch.label)
            
            acc = binary_accuracy(predictions, batch.label)

            epoch_loss += loss.item()
            epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)

We'll also create a function to tell us how long an epoch takes to compare training times between models.

In [ ]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

We then train the model through multiple epochs, an epoch being a complete pass through all examples in the training and validation sets.

At each epoch, if the validation loss is the best we have seen so far, we'll save the parameters of the model and then after training has finished we'll use that model on the test set.

In [ ]:
N_EPOCHS = 5

best_valid_loss = float('inf')

for epoch in range(N_EPOCHS):

    start_time = time.time()
    
    train_loss, train_acc = train(model, train_iterator, optimizer, criterion)
    valid_loss, valid_acc = evaluate(model, valid_iterator, criterion)
    
    end_time = time.time()

    epoch_mins, epoch_secs = epoch_time(start_time, end_time)
    
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), 'tut1-model.pt')
    
    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(f'\tTrain Loss: {train_loss:.3f} | Train Acc: {train_acc*100:.2f}%')
    print(f'\t Val. Loss: {valid_loss:.3f} |  Val. Acc: {valid_acc*100:.2f}%')

Epoch: 01 | Epoch Time: 10m 45s
	Train Loss: 0.694 | Train Acc: 49.16%
	 Val. Loss: 0.696 |  Val. Acc: 50.36%
Epoch: 02 | Epoch Time: 10m 40s
	Train Loss: 0.693 | Train Acc: 50.31%
	 Val. Loss: 0.695 |  Val. Acc: 49.04%
Epoch: 03 | Epoch Time: 10m 39s
	Train Loss: 0.693 | Train Acc: 50.50%
	 Val. Loss: 0.696 |  Val. Acc: 50.02%
Epoch: 04 | Epoch Time: 10m 43s
	Train Loss: 0.693 | Train Acc: 50.63%
	 Val. Loss: 0.696 |  Val. Acc: 50.15%
Epoch: 05 | Epoch Time: 10m 38s
	Train Loss: 0.693 | Train Acc: 50.37%
	 Val. Loss: 0.695 |  Val. Acc: 49.17%


In [ ]:
model.load_state_dict(torch.load('tut1-model.pt'))

test_loss, test_acc = evaluate(model, test_iterator, criterion)

print(f'Test Loss: {test_loss:.3f} | Test Acc: {test_acc*100:.2f}%')

Test Loss: 0.706 | Test Acc: 47.98%


### References

https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/1%20-%20Simple%20Sentiment%20Analysis.ipynb

http://anie.me/On-Torchtext/

https://github.com/spro/practical-pytorch